File: java/src/main/java/cn/az/code/medium/HouseRobber.java
package cn.az.code.medium;

import cn.az.code.common.TreeNode;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * house robber
 * @author az
 * @date 2020/4/16
 */
public class HouseRobber {

    public static void main(String[] args) {

    }

    /**
     * put map out(for recursive)
     */
    Map<TreeNode, Integer> memo = new HashMap<>(16);

    public int rob(TreeNode root) {
        if (Objects.isNull(root)) {
            return 0;
        }
        if (memo.containsKey(root)) {
            return memo.get(root);
        }

        int doIt = root.val
                + (Objects.isNull(root.left) ? 0 : rob(root.left.right))
                + (Objects.isNull(root.right) ? 0 : rob(root.right.left));

        int notDoIt = rob(root.left) + rob(root.right);

        int res = Math.max(doIt, notDoIt);
        memo.put(root, res);
        return res;
    }

    public int robDp(TreeNode root) {
        int[] res = dp(root);
        return Math.max(res[0], res[1]);
    }

    public int[] dp(TreeNode root) {
        if (Objects.isNull(root)) {
            return new int[]{0, 0};
        }
        int[] left = dp(root.left);
        int[] right = dp(root.right);

        int rob = root.val + left[0] + right[0];
        int notRob = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);

        return new int[]{notRob, rob};
    }
}

File: src/dp/house_robber.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution;

int main()
{
    auto s = new Solution();
    auto data = vector({1, 2, 3, 1});
    auto ret = s->rob(data);
    cout << "ret" << ret << endl;
    return 0;
}

class Solution
{
private:
    vector<int> memo;

public:
    int rob(vector<int> &nums)
    {
        this->memo = vector<int>(nums.size(), -1);
        return this->dp(nums, 0);
    }

    int dp(vector<int> &nums, int start)
    {
        if (start >= nums.size())
        {
            return 0;
        }
        if (memo[start] != -1)
        {
            return memo[start];
        }
        int res = max(this->dp(nums, start + 1), nums[start] + this->dp(nums, start + 2));
        memo[start] = res;
        return res;
    }
};